### PR TITLE
Bounce improvements

### PR DESCRIPTION
**Pull request by StandingPadAnimations**
These changes are being added to the optimizer (once I finish them up):
* Volumetric stepping rate modification 
* Automatic scrambling distance optimization
* Better glossy and transmissive optimization (in a different way compared to how I planned on doing it)

### CODE REVIEW

**Comment by StandingPadAnimations** [comment]
The improvements are pretty much finished. All that is needed is a code review before merging

**Comment by StandingPadAnimations** [comment]
@TheDuckCow just requesting a code review before merging

**Comment by TheDuckCow** [comment]
Thanks for flagging - taking a look but might not finish in one sitting. Will do by (my) tomorrow morning latest.

File: MCprep_addon/optimize_scene.py [hunk 1]
```diff
@@ -31,6 +31,13 @@
 CMP_BOUNCES = MIN_BOUNCES * 2
 MAX_FILTER_GLOSSY = 1.0
 MAX_STEPS = 200
+MIN_SCRAMBLING_MULTIPLIER = 0.35
```
Comment (TheDuckCow) [inline MCprep_addon/optimize_scene.py:34]: For these any any other above constants, if you have any one-line comment to add to why these values are set, would be greatly helpful.

File: MCprep_addon/optimize_scene.py [hunk 2]
```diff
@@ -142,12 +159,9 @@ def execute(self, context):
 
 		# Optimizer Settings.
 		Quality = scn_props.quality_vs_speed
-
-		# Render engine settings.
-		# TODO: Add better volumetric optimizations by checking volumetric materials and enabling certain features that benifit the scene (such as homogeneous)
-		if scn_props.volumetric_bool:
-			Samples += 50
-			Volume = 2
+		UseScrambling = scn_props.scrambling_unsafe
```
Comment (TheDuckCow) [inline MCprep_addon/optimize_scene.py:162]: nit about variable naming: `uses_scrambling` here and below (there's some variable names from your prior merge too that I didn't get to updating)

File: MCprep_addon/optimize_scene.py [hunk 3]
```diff
@@ -157,17 +171,22 @@ def execute(self, context):
...
 			FilterGlossy = MAX_FILTER_GLOSSY
 			MaxSteps = MAX_STEPS // 2 # TODO: Add better volumetric optimizations
 
 		# Compute device.
 		if cycles_compute_device_type == "NONE":
+			if UseScrambling:
+				UseScrambling = False # Automatic scrambling distance is useless on CPU
+				PreviewScrambling = False
+				ScramblingMultiplier = 1
+				print("Disabling Automatic Scrambling Distance as it is useless for CPU rendering!")
```
Comment (TheDuckCow) [inline MCprep_addon/optimize_scene.py:189]: Does it harm rendering keeping it on? Something to consider is someone who works on their file in one machine, but then renders on another (I do this often). Obviously they won't get all the best optimizations unless they ran optimize again on the other machine, but not always feasible for a render farm. If there are some settings that have no harm to CPU but positively affect GPU, I'd just always enable it.

File: MCprep_addon/optimize_scene.py [hunk 4]
```diff
@@ -218,10 +237,63 @@ def execute(self, context):
...
+				mat_type = "reflective"
 			if generate.checklist(canon, "glass"):
 				Transmissive += 1
+				mat_type = "glass"
+
+			if mat.use_nodes:
+				nodes = mat.node_tree.nodes
+				for node in nodes:
+					if node.bl_idname == "ShaderNodeVolumeScatter" or \
+						node.bl_idname == "ShaderNodeVolumeAbsorption" or \
```
Comment (TheDuckCow) [inline MCprep_addon/optimize_scene.py:254]: I'd rephase this to be easier to read, right now it's hard to see where the if ends (plus it's not advised to need to use `\` for line continuance, and multi line if's are just icky in general imo):

```python
vol_nodes = [
    "ShaderNodeVolumeScatter",
    "ShaderNodeVolumeAbsorption",
    "ShaderNodeVolumePrincipled"
]
if node.bl_idname in vol_nodes:
    ...
```

File: MCprep_addon/optimize_scene.py [hunk 5]
```diff
@@ -218,10 +237,63 @@ def execute(self, context):
...
+						node.bl_idname == "ShaderNodeVolumePrincipled":
```
Comment (TheDuckCow) [inline MCprep_addon/optimize_scene.py:255]: Am I reading this right, is this changing the type of the node? Might this have other odd side effects?

File: MCprep_addon/optimize_scene.py [hunk 6]
```diff
@@ -218,10 +237,63 @@ def execute(self, context):
...
+						if ScramblingMultiplier >= CMP_SCRAMBLING_MULTIPLIER: # at this point, it's worthless to keep it enabled 
+							UseScrambling = False
+							PreviewScrambling = False
+							ScramblingMultiplier = 1.0
+					# Not the best check, but better then nothing
+					elif node.bl_idname == "ShaderNodeBsdfPrincipled":
+						if mat_type == "reflective":
+							roughness_socket = node.inputs["Roughness"]
+							if not roughness_socket.is_linked and roughness_socket.default_value >= 0.2:
+								Glossy = Glossy - 1 if Glossy > 1 else Glossy
```
Comment (TheDuckCow) [inline MCprep_addon/optimize_scene.py:281]: nit alt phrasing idea that is easier to read (also note the renaming to lower case... I can also make that change post merge if you don't want to spend time renaming vars):

```
glossy -= 1 if glossy > 1 else 0
```

File: MCprep_addon/optimize_scene.py [hunk 7]
```diff
@@ -218,10 +237,63 @@ def execute(self, context):
...
 				Transmissive += 1
+				mat_type = "glass"
+
+			if mat.use_nodes:
+				nodes = mat.node_tree.nodes
+				for node in nodes:
+					if node.bl_idname == "ShaderNodeVolumeScatter" or \
+						node.bl_idname == "ShaderNodeVolumeAbsorption" or \
+						node.bl_idname == "ShaderNodeVolumePrincipled":
+						density_socket = node.inputs["Density"] # Grab the density
```
Comment (TheDuckCow) [inline MCprep_addon/optimize_scene.py:256]: Let's move this indented level and below into another function if possible, it's getting way too indented here.

File: README.md [hunk 8]
```diff
@@ -452,6 +452,30 @@ To add your own rigs to the Mob Spawner:
   - Alpha Threshold: At this level or lower of alpha values (0.0-1.0), delete the face from the resulting mesh.
   - Use transparent pixels: If enabled, will setup the material so that transparent pixels will appear transparent in the render, othwise will be solid.
 
+### Cycles Optimizer
```
Comment (TheDuckCow) [inline README.md:455]: Thanks for adding docs!!

File: MCprep_addon/util.py [hunk 9]
```diff
@@ -184,8 +183,17 @@ def obj_copy(base, context=None, vertex_groups=True, modifiers=True):
 	return new_ob
 
 
-def min_bv(version):
+def max_bv(version, *, up_to=None):
```
Comment (TheDuckCow) [inline MCprep_addon/util.py:186]: Per comment in other PR, isn't this about the equiv to inverting the output of min_bv? I'd rather not add a new function.

File: MCprep_addon/util.py [hunk 10]
```diff
@@ -184,8 +183,17 @@ def obj_copy(base, context=None, vertex_groups=True, modifiers=True):
...
 	if hasattr(bpy.app, "version"):
+		if up_to is not None:
+			return bpy.app.version < version
+		return bpy.app.version <= version
+
+
+def min_bv(version, *, above=None):
```
Comment (TheDuckCow) [inline MCprep_addon/util.py:193]: Would rename `above=None` to be `inclusive=False` and inver the if/else below.

**Review by TheDuckCow — CHANGES_REQUESTED** [review]
Mostly minor changes (other than variable renaming to follow snake_case, but I'm happy to do that post merge myself since it's nit picky)

**Comment by StandingPadAnimations** [comment]
Everything should be improved now. I'll merge this PR now

**Pull request merged**